### ISSUE TITLE
Inconsistent behavior observed when using named volume nested within bind-mounted volume

### ISSUE THREAD

**Issue by PacoHuerta** [docker/compose#7196]
## Description of the issue
```
version: '3.7'

volumes:
    node_modules:

services:
  web:
    image: node:latest
    volumes:
      - '.:/usr/src/docker_test/'
      - 'node_modules:/usr/src/docker_test/node_modules'
    working_dir: /usr/src/docker_test
    command: npm start

    ports:
      - '3000:3000'
```
When a named volume is mounted within a bind-mounted directory as above, the contents of the `docker_test/node_modules` directory within the container, which should point to named volume `node_modules`, will initially be separate from the node_modules directory that is within the bind-mounted directory on the host machine (`docker_test/node_modules` within host machine). This is expected; however, after a few updates to project files (described below in steps to reproduce), the contents of `docker_test/node_modules` within the container will actually be the same as that which is within the node_modules folder within the bind-mounted folder on the host machine (The contents of `docker_test/node_modules` in the host machine will eventually replace the contents of `docker_test/node_modules` within the container).

## Context information (for bug reports)

**Output of `docker-compose version`**
```
docker-compose version 1.25.2, build 698e2846
docker-py version: 4.1.0
CPython version: 3.7.5
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019
```

**Output of `docker version`**
```
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:34 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker-compose config`**
```
services:
  web:
    command: npm start
    image: node:latest
    ports:
    - published: 3000
      target: 3000
    volumes:
    - /Users/blsz/src/docker_test/docker_test:/usr/src/docker_test:rw
    - node_modules:/usr/src/docker_test/node_modules:rw
    working_dir: /usr/src/docker_test
version: '3.7'
volumes:
  node_modules: {}
```

## Steps to reproduce the issue
I have created a toy repository that can be used to reproduce this issue. 
1. Clone toy project from https://github.com/PacoHuerta/docker_test (simple template created with create-react-app)
2. From within the project root (docker_test), execute `docker-compose up`. This will install dependencies, as the `npm start` command has been changed to `"npm install && react-scripts start"`. This will also start the dev server.
3. Once server is running, open a new terminal window, and from within `docker_test` execute ls `node_modules`. Contents should be empty, as expected.
4. Execute `touch node_modules/zzz.txt`. `node_modules` directory of host machine should now only contain the file `zzz.txt`.
5. Execute `docker container exec -it docker_test_web_1 bash`
6. Once within the container, execute `ls node_modules`. Verify that the dependencies are listed within the container's `node_modules` directory, and that file `zzz.txt` is not to be found there.
7. execute `exit` to return to host machine
8. Open browser and navigate to `localhost:3000`. React logo should display (create-react-app landing page).
9. From host machine's `docker_test` directory, execute `vim public/index.html`.
10. Right underneath the `<noscript>` tag, insert text `this is some text right here`. Save and quit out of vim with `:wq`. The browser should refresh, and "this is some text right here" should show at the top of the page.
11. execute `docker container exec -it docker_test_web_1 bash`
12. execute `ls node_modules`

### Observed result
The `node_modules` directory within the container no longer contains the dependencies that were installed through `npm install`. Instead, the directory now holds the file `zzz.txt`, which was created in step 4.

**Note:** This may not happen on the first try. It is possible that steps 9-12 may have to be executed multiple times before the contents of the `node_modules` directory from the host machine are copied over to the contents of the `node_modules` directory within the container. If steps 9-12 need to be repeated, just keep alternating the text changes to `public/index.html` between `this is some text right here` and `this is some text right here man`. After about 10 or 15 tries, the issue should be visible. 

The fact that one must retry steps 9-12 an indeterminate amount of times is what leads me to believe that this is a bug. The behavior is not consistent. 

**Note:** It is also possible that the issue manifests itself as soon as step 6 is executed; however, it is most likely that steps 9-12 will have to be repeated a few times before the issue is seen.
### Expected result
It is expected that the `node_modules` directory within the container contains the dependencies installed through `npm install` and that it does not contain the file `zzz.txt`. On the host machine, it is expected that `docker_test/node_modules` contains only the file `zzz.txt`.
### Stacktrace / full error message
No stacktrace is produced.

## Additional information

MacOS Mojave 10.14.2, Docker for Mac

**Comment by PacoHuerta** [comment]
**Update**

After executing 
```
sudo screen ~/Library/Containers/com.docker.docker/Data/vms/0/tty
cd /var/lib/docker/volumes/docker_test_node_modules/_data
ls
```

I have confirmed that the volume data is healthy. That is, the volume data still contains the node module dependencies that were initially verified in step 6 of the issue reproduction steps. Therefore, the issue is solely that the contents of `docker_test/node_modules` within the container get replaced with the contents of `docker_test/node_modules` within the host machine. The volume data is not being overwritten. The cotaniner is simply losing its connection to the named volume and is instead bound to the contents of `docker_test/node_modules` within the host machine.

**Comment by coox** [comment]
@PacoHuerta —

It would be interesting to know the output of:
- `df -h` within the container (showing the state of mounted disks)
- the `Mounts` section of a `docker inspect` on your container

We **may** have observed a similar problem with a Ubuntu-based container today, but on a Windows host, while a macOS host consuming the same docker-compose.yaml file behaves as expected.

Our base image is derived from [heroku/heroku:18](https://hub.docker.com/r/heroku/heroku/).

Our service and volumes declaration is essentially the same, the relevant parts reading:

```yaml
services:
  app:
    volumes:
      - ./${APP_PATH}/:/app/:cached
      - app-node-modules:/app/node_modules/

volumes:
  app-node-modules:
```

In our case, investigation led us to notice that upon disk inspection within the container:

- the `/app` volume consistently appears to be mounted when on both macOS host and Windows host
- the `/app/node_modules` volume consistently appears to be mounted when on macOS host
- the `/app/node_modules` volume consistently appears **not** to be mounted when on Windows host

```
# df output on Ubuntu container: Notice that /app/node_modules is showing as intended
# Host: macOS Catalina 10.15.2, Docker 19.03.5, docker-compose 1.25.2

user@0cefa9952c23:/app$ df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          59G   11G   46G  19% /
tmpfs            64M     0   64M   0% /dev
tmpfs           3.0G     0  3.0G   0% /sys/fs/cgroup
overlay         3.0G  700K  3.0G   1% /sbin/docker-init
osxfs           466G   55G  397G  12% /app
/dev/sda1        59G   11G   46G  19% /app/node_modules
shm              64M     0   64M   0% /dev/shm
tmpfs           3.0G     0  3.0G   0% /proc/acpi
tmpfs           3.0G     0  3.0G   0% /sys/firmware
```

```
# df output on Ubuntu container: Notice that /app/node_modules is *not* showing as intended
# Host: Windows 10 , Docker 19.03.5, docker-compose 1.24.1 (1.25.2 shows similar results)

user@c59a084e9e80:/app$ df -h
overlay          59G   12G   45G  21% /
tmpfs            64M     0   64M   0% /dev
tmpfs           3.0G     0  3.0G   0% /sys/fs/cgroup
overlay         3.0G  872K  3.0G   1% /sbin/docker-init
grpcfuse        477G   78G  400G  17% /app
/dev/sda1        59G   12G   45G  21% /etc/hosts
shm              64M     0   64M   0% /dev/shm
tmpfs           3.0G     0  3.0G   0% /proc/acpi
tmpfs           3.0G     0  3.0G   0% /sys/firmware
```

When inspecting the `Mounts` section of `docker inspect`, we notice that:

- the mount with destination `/app/node_modules` consistently appears **after** the mount with destination `/app` when on macOS
- the mount with destination `/app/node_modules` consistently appears **before** the mount with destination `/app` when on Windows

```
# docker inspect output: Notice that /app/node_modules appears *after* /app
# Host: macOS Catalina 10.15.2, Docker 19.03.5, docker-compose 1.25.2

{
  "Mounts": [
    {
      "Type": "bind",
      "Source": "/Path/To/app",
      "Destination": "/app",
      "Mode": "cached",
      "RW": true,
      "Propagation": "rprivate"
    },
    {
      "Type": "volume",
      "Name": "localdev_app-node-modules",
      "Source": "/var/lib/docker/volumes/localdev_app-node-modules/_data",
      "Destination": "/app/node_modules",
      "Driver": "local",
      "Mode": "rw",
      "RW": true,
      "Propagation": ""
    }
  ]
}
```

```
# docker inspect output: Notice that /app/node_modules appears *before* /app
# Host: Windows 10 , Docker 19.03.5, docker-compose 1.24.1 (1.25.2 shows similar results)

 {
  "Mounts": [
    {
      "Type": "volume",
      "Name": "localdev_app-node-modules",
      "Source": "/var/lib/docker/volumes/localdev_app-node-modules/_data",
      "Destination": "/app/node_modules",
      "Driver": "local",
      "Mode": "rw",
      "RW": true,
      "Propagation": ""
    },
    {
      "Type": "bind",
      "Source": "/host_mnt/c/Path/To/app",
      "Destination": "/app",
      "Mode": "rw",
      "RW": true,
      "Propagation": "rprivate"
    }
  ]
}
```

Could it be that the mounts are not done in the order specified in the docker-compose file under certain circumstances?

**Comment by PacoHuerta** [comment]
@coox -- 

**Output of `df -h` from within the container once issue manifests itself**
```
Filesystem      Size  Used Avail Use% Mounted on
overlay          59G   30G   26G  54% /
tmpfs            64M     0   64M   0% /dev
tmpfs           2.0G     0  2.0G   0% /sys/fs/cgroup
/dev/sda1        59G   30G   26G  54% /etc/hosts
shm              64M     0   64M   0% /dev/shm
osxfs           466G  184G  281G  40% /usr/src/docker_test
tmpfs           2.0G     0  2.0G   0% /proc/acpi
tmpfs           2.0G     0  2.0G   0% /sys/firmware
```

The above output is very similar to the output that you are getting from within the container on your Windows host machine. I am not showing the `node_modules` volume as mounted once the issue manifests itself. **Note:** I checked the output of `df -h` before the issue occurred, and the output is the same as above. I do not see a line that is specific to `/usr/src/docker_test/node_modules`. I only see the line specific to `/usr/src/docker_test`. Not sure why I am not seeing the line specific to the named volume.

**Mounts section from docker inspect**
```
"Mounts": [
            {
                "Type": "volume",
                "Name": "docker_test_node_modules",
                "Source": "/var/lib/docker/volumes/docker_test_node_modules/_data",
                "Destination": "/usr/src/docker_test/node_modules",
                "Driver": "local",
                "Mode": "rw",
                "RW": true,
                "Propagation": ""
            },
            {
                "Type": "bind",
                "Source": "/Users/paco.huerta/src/docker_test",
                "Destination": "/usr/src/docker_test",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            }
        ]
```
I am also seeing the mounts listed in the same order as the mounts of the container that is giving you issues; however, I am not so sure that the order is the problem, as bind mounts and volumes are mounted in lexicographical order according to [this issue thread.](https://github.com/moby/moby/issues/26051#issuecomment-242847347) Those comments are a bit dated, but I don't see why the mount order would change. I am beginning to think that this issue should be opened somewhere else, perhaps on Moby, as I am not sure if docker-compose is directly at fault here.

So your `node_modules` directory always points to the named volume on your Mac then, @coox? You never lose the binding to the named volume? I have now been able to reproduce this issue on two different Mac machines, both on OSX Mojave. The second machine is on Mojave 10.14.6. It doesn't take very long to reproduce the issue either. I just have to repeat steps 9-12 a few times and I will see it happen.

@cpuguy83 sorry to tag you directly here, but I have read many posts that you have contributed to that have helped me gain a good understanding of the issue in this thread (like the one linked above and also [this one.](https://github.com/moby/moby/issues/19120#issuecomment-352450375) Would you happen to have any insight as to why we are experiencing this issue?

**Comment by PacoHuerta** [comment]
Closing this ticket, as I believe it is not directly caused by docker-compose. Moving discussion to: https://github.com/moby/moby/issues/40441

**Comment by coox** [comment]
For what it’s worth, the issue that I [described earlier](https://github.com/docker/compose/issues/7196#issuecomment-579784863) was Windows-specific, and has since been fixed in version 2.2.0.3, per https://github.com/docker/for-win/issues/5540.
